### PR TITLE
build: Improved hotswap behavior for the IDE's JVM when executed via the `runIde` Gradle task.

### PR DESCRIPTION
**Pull request by SCWells72**
Pretty much what it says in the title. This allows dynamic updates to the running IDE JVM instance for most changes. Changes to class constant values and significant restructuring of Java/Kotlin types will require a JVM restart, but most other changes -- including simple method extractions and such -- will not.

Once you did some Java code changes, you need to build project 

image

(ex: Ctrl+F9 on Windows) and your changes is applied without restarting the IntelliJ! This feature works only for IJ > 2024.1

### CODE REVIEW

**Comment by fbricon** [comment]
Please rebase your PR against master, we're keeping a linear git history. Merge commits are likely messing up with the conventional commits check

**Comment by angelozerr** [comment]
I have tried the PR, but I cannot start the plugin:

`Caused by: java.lang.VerifyError: Catch type is not a subclass of Throwable in exception handler 57`

```
Execution failed for task ':runIde'.
> Process 'command 'C:\Users\XXXX\.gradle\caches\modules-2\files-2.1\com.jetbrains\jbre\jbr_jcef-17.0.7-windows-x64-b1000.6\extracted\jbr_jcef-17.0.7-windows-x64-b1000.6\bin\java.exe'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
For more on this, please refer to https://docs.gradle.org/8.5/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
BUILD FAILED in 53s
10 actionable tasks: 3 executed, 7 up-to-date
Configuration cache entry stored.


```

**Comment by angelozerr** [comment]
@SCWells72 have you installed some external plugin? In my case I can start runIde (see my previous comment).

File: build.gradle.kts [hunk 1]
```diff
@@ -172,7 +172,21 @@ tasks {
         //TODO inject changelog into plugin.xml change-notes
     }
 
+    // TODO: Once LSP4IJ no longer supports 2023.*, the hotswap JVM args can be added unconditionally
+    fun supportsEnhancedClassRedefinition(): Boolean {
+        val platformVersion = properties("platformVersion").getOrNull()
+        return when (platformVersion) {
+            // Versions before 2024.1 don't support enhanced class redefinition
+            "2023.2", "2023.3" -> false
```
Comment (SCWells72) [inline build.gradle.kts:180]: I've explicitly verified the correct behavior in 2023.2, 2023.3, 2024.1, 2024.2, and 2024.3. By "correct behavior", I mean that `runIde` actually does run an IDE instance of the specified `platformVersion`, and once started, only simple classloader hotswaps are supported on 2023.* and enhanced classloader hotswaps are supported on 2024.*.

**Comment by angelozerr** [comment]
@SCWells72 it works like a charm. Thanks so much for this great PR. It will change my life :)

File: build.gradle.kts (hunk 1 above)
Comment (adietish) [reply build.gradle.kts:180]: that's great news, thanks for sharing!

**Pull request merged**